### ISSUE TITLE
Doubled DirectDebit transaction processed

### ISSUE THREAD

**Issue by sgladstone** [iATSPayments/com.iatspayments.civicrm#70]
For ALL of the monthly recurring contributions in a production environment, the transaction for March was incorrectly processed twice. ( on March 7).   (Previous months were fine). Although the original schedule was for the 11th of each month (for this example contact), yet starting in Jan. 2015, transactions were processed on the 7th of each month. (This recurring contribution was originally set up in July 2014)

Screen prints of one of the impacted contacts is attached. 
![contribs](https://cloud.githubusercontent.com/assets/2640984/6620315/63964bec-c8a6-11e4-97e2-2a0f04fc0cd5.png)
![recur_schedule](https://cloud.githubusercontent.com/assets/2640984/6620318/69651206-c8a6-11e4-917b-d7eb91510148.png)

**Comment by KarinG** [comment]
Ok that's not good

1.3.1 (latest release) came out Dec. 15 - 2014 - why would it work properly and then start to double process in March?

Or are you suggesting CiviCRM core 4.4.12 -> 4.4.13 has introduced something that is causing this issue? I will start a new test-grid for ACH on 4.4.13 today - since it's Th it will take till early next week to Complete such a recurring ACH/test series.

**Comment by sgladstone** [comment]
Karin - The version of CiviCRM in this situation is 4.4.7, and has not changed in the last 2 months.

**Comment by KarinG** [comment]
So things went bananas in March but neither CiviCRM core nor the iATS extension have changed over the past couple months. Does that not mean then that you must have done something specific to this install - re: code/extensions/manipulating tables/cron jobs/restore from backups? I'm not sure what you're expecting us to look at here.

**Comment by sgladstone** [comment]
The only changes have been:  

- Non-iATS extensions have been updated ( not other payment processors)
- The URL changed (old URL:  "www2.something.org"  new URL "something.org" )
- The iATS extension was updated in January due to a previous issue.

**Comment by sgladstone** [comment]
My hunch is that it was a race condition, caused by the CiviCRM cron job that initiates all the scheduled jobs getting called twice within a seconds.

**Comment by adixon** [comment]
Yes, since the contribution times are the same, I think you're correct. I wouldn't expect this to happen for a healthy install, but perhaps some kind of locking mechanism would be good to prevent this, ideally as part of civicrm core since it would likely affect other scripts as well.

Do you see anything that would be useful to do for the iats extension?

**Comment by sgladstone** [comment]
I would suggest creating a "in process" flag somewhere (such as in a database table) with the key being the contribution_recur ID and the current date.  So if the iATS scheduled job is called, it first checks that "in process" flag is blank/null.  If its empty, the job would set "in process=true" , then proceed to start creating a new transaction for that recurring ID schedule.  When that contribution is done, then the job would set "in process=false".     

So if the job is called twice in the same timeframe, the second job would see the "in process=true" flag, and not do anything for that recurring ID. 

There is an article about dealing with locks in PHP at: https://locallost.net/?p=1091

**Comment by adixon** [comment]
Yes, but wouldn't this make more sense to put in core? Or are there valid cases where you'd want to be running parallel cron jobs?

**Comment by sgladstone** [comment]
I think there are valid cases for other (non-iATS) scheduled jobs to run in parallel - like for the mass mail scheduled job where people may by trying to send thousands of mass email and want it to go in parallel to expedite the delivery.

**Comment by adixon** [comment]
Actually, that's probably a bad example, you want your mail job to manage it's own parallel processing I think, but perhaps there are some jobs that are going to be less prone to issues with parallel invocations.

I wouldn't want to be inventing this kind of wheel in isolation though, so let's move this over to core - which should either provide a configuration of the job that does locking, or at least a common library for jobs that want to implement locking.

Although setting up locking isn't too hard, it creates it's own issues, for example when jobs get interrupted and not allowed to clear out their own locks, you need something to time out those locks...

Conclusion: good issue, but out of scope for a single extension.

**Comment by sgladstone** [comment]
You should update the README cautioning people that doubled transactions are possible with this extension.

**Comment by adixon** [comment]
I've done some research and see that a library for locking for jobs is already available, and implemented for the mail case. I've implemented it for the recurring contributions job now, as well.

It'll be available in the next release, or you can use this commit to put it into your existing install and test it.

https://github.com/iATSPayments/com.iatspayments.civicrm/commit/1660559607080decd9096e4053ca34e3775b0388

**Comment by sgladstone** [comment]
Were you able to reproduce the issue in a test environment? And then use this patch to verify the the change solves the issue?     Also, when is the next release scheduled?

**Comment by adixon** [comment]
New release 1.3.2 today includes the locking.
